elements/pfe-avatar/src/pfe-avatar.js

```diff
@@ -0,0 +1,352 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+import PFElement from "../pfelement/pfelement.js";
+import { hash } from "./djb-hash.js";
+import { hsl2rgb, rgb2hsl } from "./hslrgb.js";
+
+class PfeAvatar extends PFElement {
+  static get tag() {
+    return "pfe-avatar";
+  }
+
+  get templateUrl() {
+    return "pfe-avatar.html";
+  }
+
+  get styleUrl() {
+    return "pfe-avatar.scss";
+  }
+
+  static get observedAttributes() {
+    return ["name", "pattern", "src", "shape"];
+  }
+
+  static get patterns() {
+    return {
+      triangles: "triangles",
+      squares: "squares"
+    };
+  }
+
+  static get defaultColors() {
+    return "#3B0083 #f0ab00 #007a87 #00b9e4 #92d400";
+  }
+
+  get name() {
+    return this.getAttribute("name");
+  }
+
+  set name(val) {
+    return this.setAttribute("name", val);
+  }
+
+  get src() {
+    return this.getAttribute("src");
+  }
+
+  set src(href) {
+    return this.setAttribute("src", href);
+  }
+
+  get pattern() {
+    return this.getAttribute("pattern") || PfeAvatar.patterns.squares;
+  }
+
+  set pattern(name) {
+    if (!PfeAvatar.patterns[name]) {
+      this.log(
+        `invalid pattern "${name}", valid patterns are: ${Object.values(
+          PfeAvatar.patterns
+        )}`
+      );
+      return;
+    }
+    return this.setAttribute("pattern", name);
+  }
+
+  constructor() {
+    super(PfeAvatar);
+  }
+
+  connectedCallback() {
+    super.connectedCallback();
+
+    this._initCanvas();
+
+    this.dispatchEvent(
+      new CustomEvent(`${PfeAvatar.tag}:connected`, {
+        bubbles: false
+      })
+    );
+  }
+
+  attributeChangedCallback(attr, oldValue, newValue) {
+    super.attributeChangedCallback(...arguments);
+
+    switch (attr) {
+      case "name":
+      case "pattern":
+        break;
+      case "src":
+        break;
+
+      default:
+    }
+
+    if (this.connected) {
+      this.update();
+    } else {
+      this.addEventListener(`${PfeAvatar.tag}:connected`, () => this.update());
+    }
+  }
+
+  _initCanvas() {
+    this._canvas = this.shadowRoot.querySelector("canvas");
+    const size = this.var("--pfe-avatar--width").replace(/px$/, "");
+    this._canvas.width = size;
+    this._canvas.height = size;
+
+    this._squareSize = this._canvas.width / 8;
+    this._triangleSize = this._canvas.width / 4;
+
+    this._ctx = this._canvas.getContext("2d");
+  }
+
+  static _registerColors() {
+    this.colors = [];
+    const themeColors =
+      PFElement.var("--pfe-avatar--colors") || this.defaultColors;
```

Comment:
Should this get the variable from PFElement or would it be more accurate to get it from this context since the variable is defined on the pfe-avatar host?

Reply:
Okay, I see this is calling a function that is declared inside of PFElement.  Does this extending PFElement mean that we could do `this.var` and it would still reference `PFElement.var`?  It might be nice to do that in general if it is the same so that if PFElement name changes again (here's hoping it doesn't), we only need to update that once in the file instead of multiple times.

Reply:
Sure, that should work but it needs testing.

Reply:
👍 Just let me know if you'd like to defer this and I can open an issue to address it later

Reply:
@kylebuch8 and I discovered this same issue. I think we should defer this so we can get some sort of MVP out soon.

Reply:
I opened this issue: https://github.com/patternfly/patternfly-elements/issues/297

Reply:
Kyle and Mark just noticed the same thing, and we looked at it closer and `this.var` won't actually work here like I thought.  Because `_registerColors` is a static function, within it `this` refers to the class `PfeAvatar`, not a `<pfe-avatar>` element.

I think these are our choices:

1. get the colors from `:root` once and apply them to all avatars (simple, efficient, less flexible)
2. get the colors for each avatar on the page (flexbile, but slow since `getComputedStyles()` is slow)